File: ios/Classes/TencentLBSHostApiImpl.swift
import Flutter
import UIKit
import TencentLBS

/// Pigeon Host API 实现（Swift 协议），与腾讯定位 SDK 文档对齐。
/// 坐标系：Flutter 统一 0=GCJ02, 1=WGS84 → iOS TencentLBSLocationCoordinateTypeGCJ02=0, WGS84=1
final class TencentLBSHostApiImpl: NSObject, TencentLBSHostApi, TencentLBSLocationManagerDelegate {
    private var locationManager: TencentLBSLocationManager?
    private var flutterApi: TencentLBSFlutterApi?
    private var isListenLocationUpdates = false

    override init() {
        super.init()
    }

    func configure(options: InitOptions) throws -> Bool {
        let manager = TencentLBSLocationManager()
        manager.delegate = self
        manager.apiKey = options.apiKey
        // enableAntiMockLocation：开启反作弊检查。mockEnable=true 表示允许 Mock，故应关闭反作弊。
        manager.enableAntiMockLocation = !(options.mockEnable ?? false)

        if let ct = options.coordinateType {
            manager.coordinateType = ct == 0 ? .GCJ02 : .WGS84
        }
        if let rl = options.requestLevel {
            let level: TencentLBSRequestLevel = {
                switch rl {
                case 0: return .geo
                case 1: return .name
                case 3: return .adminName
                case 4: return .poi
                default: return .adminName
                }
            }()
            manager.requestLevel = level
        }
        locationManager = manager
        return true
    }

    func setUserAgreePrivacy(agree: Bool) throws {
        TencentLBSLocationManager.setUserAgreePrivacy(agree)
    }

    func requestLocationOnce() throws {
        locationManager?.requestLocation(completionBlock: { [weak self] location, err in
            guard let self = self else { return }
            if let e = err {
                let code = (e as NSError).code
                self.flutterApi?.onError(code: Int64(code), message: e.localizedDescription) { _ in }
            } else if let loc = location {
                self.sendLocationToFlutter(loc)
            }
        })
    }

    func startLocationUpdates(request: ContinuousLocationRequest, androidNotificationOptions: AndroidNotificationOptions?) throws {
        guard let manager = locationManager, !isListenLocationUpdates else { return }
        guard request.intervalMs >= 1000 else {
            throw FlutterError(code: "invalid-argument", message: "intervalMs 必须大于等于 1000，当前值: \(request.intervalMs)", details: nil)
        }
        isListenLocationUpdates = true
        manager.locationCallbackInterval = UInt64(request.intervalMs)
        if request.backgroundLocation == true {
            manager.allowsBackgroundLocationUpdates = true
            manager.pausesLocationUpdatesAutomatically = false
        }
        if let rl = request.requestLevel {
            let level: TencentLBSRequestLevel = {
                switch rl {
                case 0: return .geo
                case 1: return .name
                case 3: return .adminName
                case 4: return .poi
                default: return .adminName
                }
            }()
            manager.requestLevel = level
        }
        manager.startUpdatingLocation()
    }

    func updateLocationRequest(update: LocationRequestUpdate?) throws {
        guard let manager = locationManager, let update = update, isListenLocationUpdates else { return }
        var needRestart = false
        if let interval = update.intervalMs {
            guard interval >= 1000 else {
                throw FlutterError(code: "invalid-argument", message: "intervalMs 必须大于等于 1000，当前值: \(interval)", details: nil)
            }
            manager.locationCallbackInterval = UInt64(interval)
            needRestart = true
        }
        if let rl = update.requestLevel {
            let level: TencentLBSRequestLevel = {
                switch rl {
                case 0: return .geo
                case 1: return .name
                case 3: return .adminName
                case 4: return .poi
                default: return .adminName
                }
            }()
            manager.requestLevel = level
            needRestart = true
        }
        // iOS 上修改 locationCallbackInterval / requestLevel 后需重启定位才会生效。
        // 在后台时用 background task 保护，避免 stop 与 start 之间被系统挂起导致定位未重新开启。
        if needRestart {
            let isBackground = UIApplication.shared.applicationState == .background
            var taskId: UIBackgroundTaskIdentifier = .invalid
            if isBackground {
                taskId = UIApplication.shared.beginBackgroundTask(withName: "TencentLBSRestart") {
                    if Self._restartTaskId != .invalid {
                        UIApplication.shared.endBackgroundTask(Self._restartTaskId)
                        Self._restartTaskId = .invalid
                    }
                }
                Self._restartTaskId = taskId
            }

            manager.stopUpdatingLocation()
            manager.startUpdatingLocation()

            if isBackground, taskId != .invalid {
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                    if Self._restartTaskId != .invalid {
                        UIApplication.shared.endBackgroundTask(Self._restartTaskId)
                        Self._restartTaskId = .invalid
                    }
                }
            }
        }
    }

    private static var _restartTaskId: UIBackgroundTaskIdentifier = .invalid

    func stopLocationUpdates() throws {
        isListenLocationUpdates = false
        locationManager?.stopUpdatingLocation()
    }

    func setBinaryMessenger(_ messenger: FlutterBinaryMessenger) {
        flutterApi = TencentLBSFlutterApi(binaryMessenger: messenger)
    }

    // MARK: - TencentLBSLocationManagerDelegate
    func tencentLBSLocationManager(_ manager: TencentLBSLocationManager, didUpdate location: TencentLBSLocation) {
        sendLocationToFlutter(location)
    }

    func tencentLBSLocationManager(_ manager: TencentLBSLocationManager, didFailWithError error: Error) {
        let code = (error as NSError).code
        flutterApi?.onError(code: Int64(code), message: error.localizedDescription) { _ in }
    }

    private func sendLocationToFlutter(_ location: TencentLBSLocation) {
        let cl = location.location
        let provider = location.locationProvider
        let unifiedSource = Int64(provider.rawValue)
        let sourceProviderString: String? = {
            switch provider {
            case .GPS: return "gps"
            case .netWork: return "network"
            case .simulated: return "simulated"
            case .accessoryGPS: return "accessory_gps"
            case .accessoryNetwork: return "accessory_network"
            case .unkown: return nil
            @unknown default: return nil
            }
        }()
        let data = LocationData(
            code: 0,
            latitude: cl.coordinate.latitude,
            longitude: cl.coordinate.longitude,
            altitude: cl.altitude,
            accuracy: cl.horizontalAccuracy,
            horizontalAccuracy: cl.horizontalAccuracy,
            verticalAccuracy: cl.verticalAccuracy,
            speed: cl.speed >= 0 ? cl.speed : nil,
            bearing: cl.course >= 0 ? cl.course : nil,
            address: location.address,
            name: location.name,
            timeIso: nil,
            timeMs: Int64(cl.timestamp.timeIntervalSince1970 * 1000),
            sourceProvider: sourceProviderString,
            locationSource: unifiedSource
        )
        flutterApi?.onLocation(location: data) { _ in }
    }
}

File: android/src/main/kotlin/com/flutter_tencent_lbs_plugin/pigeon/TencentLBSHostApiImpl.kt
package com.flutter_tencent_lbs_plugin.pigeon

import android.content.Context
import android.os.Looper
import com.flutter_tencent_lbs_plugin.models.NotificationOptions
import com.flutter_tencent_lbs_plugin.utils.NotificationUtils
import com.tencent.map.geolocation.TencentLocation
import com.tencent.map.geolocation.TencentLocationListener
import com.tencent.map.geolocation.TencentLocationManager
import com.tencent.map.geolocation.TencentLocationManagerOptions
import com.tencent.map.geolocation.TencentLocationRequest

/**
 * Pigeon Host API 实现，与腾讯定位 SDK 文档对齐。
 * 坐标系：Flutter 统一 0=GCJ02, 1=WGS84 → Android COORDINATE_TYPE_GCJ02=1, COORDINATE_TYPE_WGS84=0
 */
class TencentLBSHostApiImpl(
    private val context: Context,
    private val flutterApi: TencentLBSFlutterApi,
) : TencentLBSHostApi, TencentLocationListener {

    private var locationManager: TencentLocationManager? = null
    private var tencentLocationRequest: TencentLocationRequest? = null
    private var isListeningLocationUpdates = false

    override fun configure(options: InitOptions): Boolean {
        TencentLocationManagerOptions.setKey(options.apiKey)
        locationManager = TencentLocationManager.getInstance(context)

        // 坐标系：Flutter 0=GCJ02 -> Android 1, Flutter 1=WGS84 -> Android 0
        val coordType = when (options.coordinateType?.toInt()) {
            0 -> TencentLocationManager.COORDINATE_TYPE_GCJ02
            1 -> TencentLocationManager.COORDINATE_TYPE_WGS84
            else -> TencentLocationManager.COORDINATE_TYPE_WGS84
        }
        locationManager?.coordinateType = coordType
        locationManager?.setMockEnable(options.mockEnable ?: false)

        val request = TencentLocationRequest.create()
        request.requestLevel = (options.requestLevel ?: TencentLocationRequest.REQUEST_LEVEL_ADMIN_AREA).toInt()
        request.locMode = (options.locMode ?: TencentLocationRequest.HIGH_ACCURACY_MODE).toInt()
        request.isAllowGPS = options.allowGps ?: true
        request.isIndoorLocationMode = options.indoorLocationMode ?: false
        request.isGpsFirst = options.gpsFirst ?: false
        request.gpsFirstTimeOut = (options.gpsFirstTimeOutMs ?: 5000).toInt().coerceIn(0, 60000)
        tencentLocationRequest = request
        return true
    }

    override fun setUserAgreePrivacy(agree: Boolean) {
        TencentLocationManager.setUserAgreePrivacy(agree)
    }

    override fun requestLocationOnce() {
        val manager = locationManager ?: return
        val request = tencentLocationRequest ?: return
        val errorCode = manager.requestSingleFreshLocation(request, this, Looper.getMainLooper())
        if (errorCode != TencentLocation.ERROR_OK) {
            flutterApi.onError(errorCode.toLong(), "requestSingleFreshLocation failed with error code $errorCode") {}
        }
    }

    override fun startLocationUpdates(
        request: ContinuousLocationRequest,
        androidNotificationOptions: AndroidNotificationOptions?,
    ) {
        val manager = locationManager ?: return
        val req = tencentLocationRequest ?: return
        if (isListeningLocationUpdates) return
        // 后台定位必须提供通知配置且 id > 0，否则无法启用前台服务
        if (request.backgroundLocation == true) {
            if (androidNotificationOptions == null || androidNotificationOptions.id.toInt() <= 0) {
                flutterApi.onError(-1L, "后台定位需要 androidNotificationOptions 且 id > 0") {}
                return
            }
        }
        isListeningLocationUpdates = true

        val intervalError = validateIntervalMs(request.intervalMs)
        if (intervalError != null) {
            isListeningLocationUpdates = false
            flutterApi.onError(-1L, intervalError) {}
            return
        }
        req.interval = request.intervalMs
        applyCommonRequestOptions(
            req = req,
            requestLevel = request.requestLevel,
            locMode = request.locMode,
            allowGps = request.allowGps,
            allowCache = request.allowCache,
            gpsFirst = request.gpsFirst,
            gpsFirstTimeOutMs = request.gpsFirstTimeOutMs,
        )

        if (request.backgroundLocation == true && androidNotificationOptions != null) {
            val notifId = androidNotificationOptions.id.toInt()
            if (notifId > 0) {
                val opts = toNotificationOptions(androidNotificationOptions)
                manager.enableForegroundLocation(notifId, NotificationUtils.buildNotification(context, opts))
            }
        }
        manager.requestLocationUpdates(req, this)
    }

    override fun updateLocationRequest(update: LocationRequestUpdate?) {
        if (update == null) return
        val manager = locationManager ?: return
        val req = tencentLocationRequest ?: return
        if (!isListeningLocationUpdates) return

        val onlyInterval = update.intervalMs != null &&
            update.requestLevel == null &&
            update.locMode == null &&
            update.gpsFirstTimeOutMs == null &&
            update.allowGps == null &&
            update.allowCache == null &&
            update.gpsFirst == null

        if (onlyInterval && update.intervalMs != null && update.intervalMs!! > 0) {
            val interval = update.intervalMs!!
            val intervalError = validateIntervalMs(interval)
            if (intervalError != null) {
                flutterApi.onError(-1L, intervalError) {}
                return
            }
            manager.changeCallbackInterval(interval)
            req.interval = interval
            return
        }

        update.intervalMs?.let {
            val intervalError = validateIntervalMs(it)
            if (intervalError != null) {
                flutterApi.onError(-1L, intervalError) {}
                return
            }
            req.interval = it
        }
        applyCommonRequestOptions(
            req = req,
            requestLevel = update.requestLevel,
            locMode = update.locMode,
            allowGps = update.allowGps,
            allowCache = update.allowCache,
            gpsFirst = update.gpsFirst,
            gpsFirstTimeOutMs = update.gpsFirstTimeOutMs,
        )

        manager.removeUpdates(this)
        manager.requestLocationUpdates(req, this)
    }

    override fun stopLocationUpdates() {
        isListeningLocationUpdates = false
        locationManager?.disableForegroundLocation(true)
        locationManager?.removeUpdates(this)
    }

    override fun onLocationChanged(location: TencentLocation?, error: Int, reason: String?) {
        if (error == TencentLocation.ERROR_OK && location != null) {
            val rawProvider = location.sourceProvider
            val unifiedSource = toUnifiedLocationSource(rawProvider)
            val acc = location.accuracy.toDouble()
            val data = LocationData(
                code = error.toLong(),
                latitude = location.latitude.toDouble(),
                longitude = location.longitude.toDouble(),
                altitude = location.altitude.toDouble(),
                accuracy = acc,
                horizontalAccuracy = acc,
                verticalAccuracy = null,
                speed = location.speed.toDouble(),
                bearing = location.bearing.toDouble().takeIf { it in 0.0..360.0 },
                address = location.address,
                name = location.name,
                timeIso = null,
                timeMs = location.time,
                sourceProvider = rawProvider,
                locationSource = unifiedSource,
            )
            flutterApi.onLocation(data) {}
        } else {
            flutterApi.onError(error.toLong(), reason ?: "定位失败") {}
        }
    }

    /**
     * 将 Android TencentLocation.getSourceProvider() 的字符串映射为与 iOS 统一的定位来源码。
     * 统一约定：-1=未知，0=GPS，1=网络，2=模拟，3=外设GPS，4=外设网络。
     */
    private fun toUnifiedLocationSource(sourceProvider: String?): Long? {
        if (sourceProvider.isNullOrEmpty()) return -1L
        return when (sourceProvider) {
            TencentLocation.GPS_PROVIDER,
            TencentLocation.BEIDOU_PROVIDER -> 0L  // 卫星/GPS
            TencentLocation.NETWORK_PROVIDER,
            TencentLocation.WIFI_PROVIDER,
            TencentLocation.CELL_PROVIDER,
            TencentLocation.COARSE_PROVIDER,
            TencentLocation.FUSED_PROVIDER -> 1L  // 网络
            TencentLocation.FAKE -> 2L            // 模拟/作弊
            else -> -1L
        }
    }

    override fun onStatusUpdate(name: String?, status: Int, desc: String?) {
        flutterApi.onStatus(LocationStatusData(name = name, status = status.toLong())) {}
    }

    private fun toNotificationOptions(p: AndroidNotificationOptions): NotificationOptions {
        return NotificationOptions(
            id = p.id.toInt(),
            channelId = p.channelId,
            channelName = p.channelName,
            contentTitle = p.notificationTitle,
            contentText = p.notificationText ?: "",
            channelDescription = p.channelDescription,
            enableVibration = p.enableVibration ?: false,
            playSound = p.playSound ?: false,
            showWhen = p.showWhen ?: false,
        )
    }

    private fun applyCommonRequestOptions(
        req: TencentLocationRequest,
        requestLevel: Long?,
        locMode: Long?,
        allowGps: Boolean?,
        allowCache: Boolean?,
        gpsFirst: Boolean?,
        gpsFirstTimeOutMs: Long?,
    ) {
        requestLevel?.toInt()?.let { req.requestLevel = it }
        locMode?.toInt()?.let { req.locMode = it }
        allowGps?.let { req.isAllowGPS = it }
        allowCache?.let { req.isAllowCache = it }
        gpsFirst?.let { req.isGpsFirst = it }
        gpsFirstTimeOutMs?.toInt()?.coerceIn(0, 60000)?.let { req.gpsFirstTimeOut = it }
    }

    /** 校验 interval（严格按指南口径）：必须 >= 1000ms。 */
    private fun validateIntervalMs(intervalMs: Long): String? {
        if (intervalMs < 1000L) {
            return "intervalMs 必须大于等于 1000，当前值: $intervalMs"
        }
        return null
    }
}
